src/uring.cc

```diff
@@ -0,0 +1,129 @@
+// Copyright 2011 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "uring.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdlib.h>
+#include <string.h>
+
+BulkStat::BulkStat(Callback callback):callback_(callback), n_pending(0) {
+  if (io_uring_queue_init(1024, &ring, 0) < 0) {
+    ring.ring_fd = -1;
+    return;
+  }
+
+  struct io_uring_probe *probe = io_uring_get_probe_ring(&ring);
+  if (!probe) {
+    Close();
+    return;
+  }
+
+  bool supported = io_uring_opcode_supported(probe, IORING_OP_STATX);
+  free(probe);
+  if (!supported) {
+    Close();
+    return;
+  }
+}
+
+void BulkStat::Queue(const char *path, void *data) {
+  if (!IsAvailable()) {
+    callback_(-1, "io_uring is not available", data);
+    return;
+  }
+
+  struct io_uring_sqe *s = io_uring_get_sqe(&ring);
+  if (!s) {
+    Wait();
+
+    s = io_uring_get_sqe(&ring);
+    if (!s) {
+      callback_(-1, "io_uring_get_sqe() failed", data);
+      return;
+    }
+  }
+
+  std::pair<void *, struct statx> value;
+  value.first = data;
+  PendingMap::iterator i = pending.insert(value).first;
+
+  io_uring_prep_statx(s, AT_FDCWD, path, AT_STATX_FORCE_SYNC,
+                      STATX_MTIME, &i->second);
+  io_uring_sqe_set_data(s, data);
+
+  ++n_pending;
+}
+
+void BulkStat::Wait() {
+  if (pending.empty())
+    return;
+
+  assert(IsAvailable());
+
+  int error = io_uring_submit_and_wait(&ring, n_pending);
+  if (error < 0) {
+    // something went terribly wrong, shouldn't happen
+    FailAll(strerror(-error));
```

Comment:
Why `-`?

Reply:
Because liburing returns negative errno values on error.

Reply:
I see. Where can I find the documentation of liburing / `io_uring_submit_and_wait`? (I've only found some higher-level documentation about it)

Reply:
Unfortunately, liburing is underdocumented, but this is some example code using that function: https://github.com/axboe/liburing/blob/fe500488190ff39716346ee1c0fe66bde300d0fb/examples/ucontext-cp.c#L245-L249